Move DashboardEmptyScreen inside DashboardViewport

majagrubic opened the pull request:
## Summary

### Intro
As part of the Dashboard-first [project](https://github.com/elastic/kibana/projects/48), we'll need to add Visualization from the dashboard. I was prototyping this flow, and there was a visible flick where the empty dashboard screen was still shown before the visualization appeared, visible from the GIF below:
![save_visualization](https://user-images.githubusercontent.com/1937956/69901347-4140f200-1378-11ea-9f0b-bb3b2ab2b1eb.gif)

### What this PR contains?
For the flick to go away, empty screen needs to be inside `DashboardViewport`. This PR refactors the code so that `DashboardEmptyScreen` is rendered inside `DashboardViewport` and gets all the relevant info it needs to render through container input.

### What this PR does not contain?
This PR does not actually add visualization from the dashboard. It does not change the layout nor the actions of the empty dashboard screen. It will be added in a subsequent PR for easier review.

### How best to review this?
I've add some comments to the relevant places below to help with the review. Some things to look out for: panel rendering, time application, fullscreen mode, report generation.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

majagrubic commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -177,6 +186,17 @@ export class DashboardAppController {
+        showLinkToVisualize: shouldShowEditHelp,
+      };
+      if (shouldShowEditHelp) {
+        emptyScreenProps.onVisualizeClick = noop;
```
noop for now.

majagrubic commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -735,6 +750,8 @@ export class DashboardAppController {
       }
     };
 
+    navActions[TopNavIds.VISUALIZE] = async () => {};
```
noop for now, this is where the new logic for the dashboard will live.

majagrubic commented on src/plugins/dashboard_embeddable_container/public/embeddable/viewport/dashboard_viewport.tsx:
```diff
@@ -82,24 +98,34 @@ export class DashboardViewport extends React.Component<DashboardViewportProps, S
-        )}
-        <DashboardGrid container={container} />
+      <div>
+        {isEmptyState ? this.renderEmptyScreen() : null}
```
The empty screen needs to be inside dashboardViewport for it to render nicely, but mustn't be inside inside `data-shared-items-container` or else its content will be included in the report generation. This logic effectively means we could have emptyScreen AND dashboard panels rendered on the screen at the same time, but if the state is always reset appropriately, this should never happen.

myasonik replied:
Instead of leaving it as a possibility to render both at the same time, should we specifically case the renderings? I wouldn't bet on state always being reset appropriately 😬 (though maybe I'm a pessimist).

majagrubic commented on src/plugins/embeddable/public/lib/containers/container.ts:
```diff
@@ -240,6 +240,7 @@ export abstract class Container<
         ...this.input.panels,
         [panelState.explicitInput.id]: panelState,
       },
+      isEmptyState: false,
```
Adding a new panel means it's not an empty state any more, so need to reset.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12722/)
* Commit: 9b97f09e01000a357c6b028172ec37a6b95d7d3c

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12729/)
* Commit: c7f1ffb54a66cfb4673b8b1297806d94cd533bbe

majagrubic commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12731/)
* Commit: c7f1ffb54a66cfb4673b8b1297806d94cd533bbe

majagrubic commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12743/)
* Commit: a34c2204d659f2dee97777ab19080e1b7ccc2e69

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12745/)
* Commit: f3e4b32084ce119121744954803a6c2519c6b67f

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12752/)
* Commit: 257be1fb4e99b283b3f6f6e5ce0ee72554db3b4d

myasonik commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_empty_screen.tsx:
```diff
@@ -18,17 +18,29 @@
+}: DashboardEmptyScreenProps) {
   const linkToVisualizeParagraph = (
-    <p className="linkToVisualizeParagraph">
+    <EuiText data-test-subj="linkToVisualizeParagraph">
```
`EuiText` only renders `div`s around everything. Can we leave the `p` tags around content?  (`<EuiText>` can wrap `<p>`.)

majagrubic replied:
I ruins the fonts if it's not `EuiText`. I am not quite where/how is the EUI stylesheet loaded. Do you know a way to keep the fonts and not have a `EuiText`?

cchaos replied:
@majagrubic I think what @myasonik is trying to say is that EuiText just provides styling for the text, but you still need to provide the text element like a paragraph `<p>` or heading `<h2>` tag. So in this instance you would keep the `EuiText` but also wrap the FormattedMessage inside a `<p>` tag.

majagrubic replied:
🤦‍♀ 🤦‍♀Let's pretend I never said that!

myasonik commented on src/plugins/dashboard_embeddable_container/public/embeddable/viewport/dashboard_viewport.tsx:
```diff
@@ -82,24 +98,34 @@ export class DashboardViewport extends React.Component<DashboardViewportProps, S
-          />
-        )}
-        <DashboardGrid container={container} />
+      <div>
```
Nit: Can this be replaced with a Fragment? It doesn't seem like you need the extra `div` wrapper for anything...

```suggestion
      <>
```

majagrubic replied:
I think the reporting didn't work properly if this wasn't a div, but will try again.

majagrubic replied:
yes, for some reason report generation fails if this is not wrapped in a `div`:
<img width="656" alt="Screenshot 2019-12-03 at 18 07 41" src="https://user-images.githubusercontent.com/1937956/70077032-e9b0b980-15f7-11ea-9eaf-e077c6dd3965.png">

majagrubic replied:
Ah, it works with the `Fragment`. But otherwise, this DOM structure is needed for reporting to be generated properly, as it needs to have an element with `data-shared-items-container` and this cannot be wrapped around empty screen, or the content of empty screen would be generated in the report as well. This will need to change when we introduce the empty panel, but I think it's good enough for now.

cchaos commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_empty_screen.tsx:
```diff
@@ -50,15 +62,15 @@ export function DashboardEmptyScreen({ showLinkToVisualize, onLinkClick }: Props
     return (
-      <p>
+      <EuiText size="m">
         <span>
```
Same this this one. You'd want to change this `span` to a `p`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13253/)
* Commit: b570569a1d3eb36bc15a2fd7e32e0239463c89bf

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13300/)
* Commit: 5e25326b86181c5cd2238ed288446f4d7c523d2e

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13462/)
* Commit: e9062560b75ef67a4bd8073909cc186a36542690

streamich commented on src/plugins/dashboard_embeddable_container/public/embeddable/grid/_dashboard_grid.scss:
```diff
@@ -34,7 +34,7 @@
   height: 100% !important; /* 1. */
   width: 100%;
-  position: absolute;
+  position: absolute !important;
```
Why is this needed?

majagrubic replied:
Same reason it's needed for the height - to override the react grid layout.

streamich commented on src/plugins/dashboard_embeddable_container/public/embeddable/viewport/dashboard_viewport.test.tsx:
```diff
@@ -33,6 +33,10 @@ import {
+import {
+  DashboardEmptyScreen,
+  DashboardEmptyScreenProps,
+} from '../../../../../legacy/core_plugins/kibana/public/dashboard/dashboard_empty_screen';
```
We cannot import from `/src/legacy` from within a New Platform plugin `/src/plugins`? Can we move this `dashboard_empty_screen` to the new platform, say `/src/plugins/dashboard` plugin?

streamich commented on src/plugins/dashboard_embeddable_container/public/embeddable/viewport/dashboard_viewport.tsx:
```diff
@@ -23,6 +23,7 @@ import { PanelState } from '../../embeddable_plugin';
 import { DashboardContainer, DashboardReactContextValue } from '../dashboard_container';
 import { DashboardGrid } from '../grid';
 import { context } from '../../../../kibana_react/public';
+import { DashboardEmptyScreen } from '../../../../../legacy/core_plugins/kibana/public/dashboard/dashboard_empty_screen';
```
Same here, we should not be importing from the legacy platform.

streamich commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -194,6 +217,8 @@ export class DashboardAppController {
         viewMode: dashboardStateManager.getViewMode(),
         panels: embeddablesMap,
         isFullScreenMode: dashboardStateManager.getFullScreenMode(),
+        isEmptyState: shouldShowEditHelp || shouldShowViewHelp,
```
Should this `isEmptyState` prop instead be named `showHelp`?

Can dashboard embeddable container infer that it is empty from the `panels` prop above?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13746/)
* Commit: b07d8d22e98592e99cac82af8433f7105bc8c6b5

majagrubic commented on test/functional/apps/dashboard/full_screen_mode.js:
```diff
@@ -74,11 +74,10 @@ export default function ({ getService, getPageObjects }) {
-      await logoButton.moveMouseTo();
-      await PageObjects.dashboard.clickExitFullScreenTextButton();
-
+    it.skip('exits when the logo button is clicked on', async () => {
```
This test starting failing after my change, but I have no idea how this was ever working in the first place. This is never entering fullscreen mode from what I can tell. There is something weird happening with the webdriver here, where it can't click on the element, because it's obstructed by something. I've already started looking into similar failures here:
https://github.com/elastic/kibana/pull/52268
and I suspect the root cause is the same. I will address this one along those.
Tested this functionality manually (entering/exiting full screen) and it works fine.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13830/)
* Commit: 57d44af77df21169575a09d0168e61c73015e74d

majagrubic commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13864/)
* Commit: 57d44af77df21169575a09d0168e61c73015e74d

majagrubic commented:
I updated the code again, since the empty state was not being reset properly, which caused the exit fullscreen button to being rendered twice, which was messing the functional test. I updated that one functional test, as I still don't understand how it was ever passing.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13893/)
* Commit: 0b4a66e6ddedc4896293c2048fcf673550a3cf94

majagrubic commented:
Jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13937/)
* Commit: b5a7cf590ec34fabe3ed44dd9cfab2aa40d7368b

majagrubic commented:
@elasticmachine merge upstream

flash1293 commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -143,6 +142,16 @@ export class DashboardAppController {
     }
     $scope.showSaveQuery = dashboardCapabilities.saveQuery as boolean;
 
+    $scope.getShouldShowEditHelp = () =>
```
Why are those put on the scope? I don't find a place in angular where they are used

majagrubic replied:
good point. they were on the scope previously, so I just left them there, but it's true that there's no good reason why they should remain there.

flash1293 commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -234,6 +257,15 @@ export class DashboardAppController {
+          dashboardContainer.renderEmpty = () => {
+            const shouldShowEditHelp = $scope.getShouldShowEditHelp();
+            const shouldShowViewHelp = $scope.getShouldShowViewHelp();
+            const isEmptyState = shouldShowEditHelp || shouldShowViewHelp;
```
It seems like those two functions are always used like this. Would it make sense to put them into a single function and just call it directly here?

majagrubic replied:
No, there is distinction between `edit` and `view` screen, it's used in the `getEmptyScreenProps` to decide which screen should be rendered.

flash1293 commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_empty_screen.tsx:
```diff
@@ -81,11 +96,19 @@ export function DashboardEmptyScreen({ showLinkToVisualize, onLinkClick }: Props
-        <h2>{constants.fillDashboardTitle}</h2>
-        {showLinkToVisualize ? addVisualizationParagraph : enterEditModeParagraph}
-      </React.Fragment>
+      <EuiPage className="dshStartScreen" restrictWidth={'36em'}>
```
Nit: could use `restrictWidth="36em"`

flash1293 commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -234,6 +257,15 @@ export class DashboardAppController {
         if (!isErrorEmbeddable(container)) {
           dashboardContainer = container;
 
+          dashboardContainer.renderEmpty = () => {
```
It feels like we are mixing responsibilities here. From my perspective, the dashboard app and the dashboard container are two separate pieces of code that serve different purposes.

Is it absolutely necessary the container knows about the empty state panel? Couldn't we make that a private detail of the app to keep the interface between the container and the app as small as possible (not injecting the renderEmpty function in a kind of dirty way into the container)? For example we could not render the container at all as long as it's empty and directly render the empty screen from the app itself in line 317:
instead of
```ts
container.render(dashboardDom);
```

we are calling

```ts
if (isEmpty) {
 reactDom.render(<DashboardEmptyScreen {...propsNStuff} />, dashboardDom);
} else {
  container.render(dashboardDom);
}
```

Now if there is an important reason the dashboard container has to know it can be empty, would it make more sense have the empty screen component living in the dashboard container itself instead of the dashboard app? In that scenario the app would just pass in a boolean prop like `showEmptyScreen` or something like this and the container would take care of the rest.

majagrubic replied:
Please see one of my comments above. Empty screen needs to live inside the `dashboardContainer`, mainly for the reporting to work properly. `data-shared-items` container needs to remain where it is. Removing that wrapper will cause the report to fail. Putting it around `DashboardEmptyScreen` means the content of the empty screen will be included in the report.

majagrubic replied:
> Now if there is an important reason the dashboard container has to know it can be empty, would it make more sense have the empty screen component living in the dashboard container itself instead of the dashboard app? In that scenario the app would just pass in a boolean prop like showEmptyScreen or something like this and the container would take care of the rest.

I am not quite sure I fully understand this question, but I think there are two things to note:
1.  The props of the empty screen are still unfortunately defined on the `dashboard_app_controller`: https://github.com/elastic/kibana/pull/51939/files#diff-b2738a4dec941f426bf944b70d057b75R185
2. `DasbhoardContainer` is receiving `renderEmpty` in "a hacky way" as it can't go through input, because all inputs must be serializable

flash1293 commented:
Not 100% sure whether introduced by this PR, but when the "Use margin between panels" setting is disabled, the light gray background color isn't shown across the full height anymore:
<img width="1601" alt="Screenshot 2019-12-09 at 11 40 19" src="https://user-images.githubusercontent.com/1508364/70429439-d038c300-1a78-11ea-8c2a-bc8eb41894cb.png">

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14130/)
* Commit: 9a128a385713ef87168809266603d74509b9de3a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14130/","number":"14130","commit":"9a128a385713ef87168809266603d74509b9de3a"}]}
PIPELINE-->

majagrubic commented:
Re: margins on empty screen, same thing on master:
<img width="1628" alt="Screenshot 2019-12-09 at 11 47 00" src="https://user-images.githubusercontent.com/1937956/70433458-b946a080-1a79-11ea-89e3-59a48cce80b7.png">

I'll fix it in one of the next iterations if that's ok.

majagrubic commented:
thank @flash1293 and @streamich for all the help in getting this done. lots more refactoring coming, stay tuned 👍

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14497/)
* Commit: c4c7ad93c5ae7127b4daa06f9fcf3b5837af5b9f

### History
* :green_heart: [Build #14130](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14130/) succeeded 9a128a385713ef87168809266603d74509b9de3a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14130/","number":"14130","commit":"9a128a385713ef87168809266603d74509b9de3a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14497/","number":"14497","commit":"c4c7ad93c5ae7127b4daa06f9fcf3b5837af5b9f"}]}
PIPELINE-->

majagrubic commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_empty_screen.tsx:
```diff
@@ -81,11 +96,19 @@ export function DashboardEmptyScreen({ showLinkToVisualize, onLinkClick }: Props
+            {showLinkToVisualize ? addVisualizationParagraph : enterEditModeParagraph}
+          </EuiPageContent>
+        </EuiPageBody>
+      </EuiPage>
```
This screen is going to change soon, so I would suggest against additional refactoring of it now.

majagrubic commented on src/plugins/dashboard_embeddable_container/public/embeddable/viewport/dashboard_viewport.tsx:
```diff
@@ -103,4 +133,13 @@ export class DashboardViewport extends React.Component<DashboardViewportProps, S
+    return (
+      <React.Fragment>
+        {this.state.isEmptyState ? this.renderEmptyScreen() : null}
+        {this.renderContainerScreen()}
```
I'm afraid this needs to stay as is for reporting to work properly for now. But I'm planning on a bigger refactor that will hopefully eliminate this hack then.

majagrubic commented:
Thanx @cchaos, will address some of your comments in a follow-up PR.
